File: data_prep_scripts/prep_source_scripts/prep_USDA_Food_And_Nutrition.py
import pandas as pd
import numpy as np
import os
import re
import requests
import time 

source = 'https://services1.arcgis.com/vdNDkVykv9vEWFX4/arcgis/rest/services/Child_Nutrition/FeatureServer'
in_path = source + '/0/query?outFields=*&where=1%3D1&f=geojson'
out_dir = '../food-data/Cleaned_data_files'
out_path = os.path.join(out_dir, 'cleaned_summer_meal_sites_api.csv')

final_cols = ['id', 'source_org', 'source_file', 'original_id', 'type', 'name', 'address', 'city', 
              'state', 'zip_code', 'county', 'location_description', 'phone', 'url', 'latitude', 
              'longitude', 'latlng_source', 'date_from', 'date_to', 'SNAP', 'WIC', 'FMNP', 
              'fresh_produce', 'food_bucks', 'free_distribution', 'open_to_spec_group', 'data_issues']

### Summer Meal Sites ###
raw_dat = requests.get(in_path).json() #pd.read_csv(in_path, encoding = 'ansi')
get_entries = []
for i in range(len(raw_dat['features'])):
    get_entries.append(pd.json_normalize(raw_dat['features'][i]['properties']))

df = pd.concat(get_entries).reset_index(drop = True)

# Filter down to Allegheny County sites only
df = df.loc[(df['Site_State'] == 'PA') & (df['Site_County'] == 'Allegheny')].reset_index()

# Assign some columns to schema fields
df['original_id'] = df['Site_ID_External']
df['name'] = df['Site_Name'].str.title() # deal with ALL CAPS
df['city'] = df['Site_City'].str.title()
df['state'] = df['Site_State']
df['zip_code'] = df['Site_Zip'].str.zfill(5)
df['address'] = df['Site_Street']
df['county'] = df['Site_County']
df['latitude'] = df['Latitude']
df['longitude'] = df['Longitude']
df['date_from'] = df['Start_Date'].apply(lambda x: time.strftime('%B %d %Y', time.localtime(x/1000)) if x != None else None)
df['date_to'] = df['End_Date'].apply(lambda x: time.strftime('%B %d %Y', time.localtime(x/1000)) if x != None else None)
                                           

## add into to location description
df["location_description"] = None 
for row in range(len(df)):
    res = ''
    if df['Site_Street2'][row] != '':
         res = res + 'Site Location Info: ' + df['Site_Street2'][row] +'; '
    if df['Service_Type'][row] != '':
        res = res + ' Service Type: ' + df['Service_Type'][row] +'; '
    if df['Site_Hours'][row] != '':
        res = res + 'Site Hours: ' + df['Site_Hours'][row]  + '; '
    if df['Comments'][row] != '':
        res = res + 'Comments: ' + df['Comments'][row] + '; '
    if df['Site_Instructions'][row] != '':
        res = res + 'Site Instructions: ' + df['Site_Instructions'][row]
    df['location_description'][row] = res
        
        
# Set some fields directly
df['source_org'] = 'Allegheny County'
df['source_file'] = source #os.path.basename(in_path)
df['type'] = "summer meal site"
df['latlng_source'] = df['source_org']

df['SNAP'] = 0
df['WIC'] = 0
df['FMNP'] = 0
# df['fresh_produce'] = 1
df['food_bucks'] = 0
df['free_distribution'] = 1
df['open_to_spec_group'] = 'children and teens 18 and younger'
df['data_issues'] = '' # start with blank field, to populate later


# Reorder and add any missing columns
df = df.reindex(columns = final_cols)

# Identify which columns we have handled
handled_cols = df.columns[~df.isna().all()] # i.e. columns that aren't all NA

# Detect and document missingness in handled columns
for col in handled_cols:
    df.loc[df[col].isna(), 'data_issues'] += '{} missing;'.format(col)

# Detect some specific data issues 
# df.loc[((df['latitude'] == 0) & (df['longitude'] == 0)), 'data_issues'] += 'latlng is (0,0);'

# Write out to CSV
df.to_csv(out_path, index = False)

File: data_prep_scripts/prep_source_scripts/prep_USDA_Food_And_Nutrition.R
## script to clean up original data sources (by fitting to schema); data provided by PFPC
## specific sources include 1) PA SNAP 2) Grow PGH

## load libs
library(tidyverse)

## data sources for new scheme
read_loc1 <- "food-data/PFPC_data_files/" ## original file source
write_loc <- "food-data/Cleaned_data_files/" ## check if this is how we want to do it

## ----------------------- read in data_model
data_mod <- readxl::read_excel("schema.xlsx", sheet = "master_table") %>% 
  filter(!str_detect(STATUS, "remove|REMOVE|eliminate")) 

## create empty dataframe according to data model; (elegant approach suggested by Connor that perserves data types)
dat0 <- data_mod %>% 
  select(field, type) %>% 
  mutate(value = case_when(type %in% c("string", "date") ~ list("a"), 
                           type %in% c("int", "float") ~ list(1),
                           type %in% "bool" ~ list(NA))) %>%
  select(-type) %>% 
  tidyr::pivot_wider(names_from = "field", values_from = "value") %>% 
  purrr::map_dfr(unlist) %>% 
  slice(-1)

#  ---------------------------------SNAP
# point to USDA API
library(httr)
library(jsonlite)
dat1 <- "https://services1.arcgis.com/RLQu0rK7h4kbsBq5/arcgis/rest/services/Store_Locations/FeatureServer/0/query?where=State%20%3D%20'PA'%20AND%20County%20%3D%20'ALLEGHENY'&outFields=*&outSR=4326&f=json"
request <- httr::GET(dat1)
object <- httr::content(request, as = "text", encoding = "UTF-8") %>% 
  jsonlite::fromJSON()

# object$features %>% str()

SNAP <- object$features$attributes %>% 
  as_tibble() #%>% mutate_all(as.character)

SNAP <- dat0 %>% 
  bind_rows(SNAP %>% filter(State %in% "PA", 
                            County %in% "ALLEGHENY") %>% 
              mutate(address = ifelse(is.na(Address_Line__2), Address, paste(Address, Address_Line__2)),
                     zip_code = as.character(Zip5),
                     original_id = as.character(ObjectId)) %>%
              select(name = Store_Name, 
                     longitude = Longitude,
                     latitude = Latitude, 
                     address, 
                     city = City,
                     state = State, 
                     zip_code,
                     county = County,
                     original_id)) %>% 
  rowwise() %>%
  mutate(source_org = "USDA Food and Nutrition Service",
         source_file = dat1,
         latlng_source = "USDA Food and Nutrition Service",
         food_bucks = NA,
         SNAP = 1,
         WIC = NA,
         FMNP = NA,
         fresh_produce = ifelse(type %in% c("farmer's market", "supermarket"), 1, NA),
         free_distribution = 0, 
         open_to_spec_group = 0,
         data_issues = "no type;no phone;no date/time info") %>% 
  ungroup()

write_csv(SNAP, paste0(write_loc, "cleaned_PA_SNAP.csv"))
rm(dat1, SNAP)


##  ----------------- growgardenpgh
dat2 <- "growpghgardens201712_readin.xlsx"
growpgh <- readxl::read_excel(paste0(read_loc1, dat2)) 

growpgh <- dat0 %>%
  bind_rows(growpgh %>%
              mutate(state = ifelse(state %in% "Pennsylvania", "PA", state),
                     original_id = as.character(grower_id),
                     zip_code = as.character(zip_code)) %>%
              select(name = urban_grower,
                     address = street_address,
                     city,
                     state,
                     zip_code,
                     latitude,
                     longitude,
                     original_id,
                     url = url)) %>%
  rowwise() %>%
  mutate(source_org = "Western Pennsylvania Regional Data Center",
         source_file = dat2,
         type = "Grow PGH Garden",
         latlng_source = "Western Pennsylvania Regional Data Center",
         food_bucks = 0, 
         SNAP = 1, 
         WIC = 0,
         FMNP = 1, 
         fresh_produce = 1, 
         free_distribution = 0, 
         open_to_spec_group = 0,
         data_issues = "no date/time;no county info") %>%
  ungroup()

write_csv(growpgh, paste0(write_loc, "cleaned_growpgh.csv"))
rm(dat2, growpgh)

###--- clean up
rm(dat0, data_mod, read_loc1, write_loc)
